src/plus/integrations/authentication/integrationAuthentication.ts

```diff
@@ -37,107 +40,262 @@ export interface IntegrationAuthenticationSessionDescriptor {
 	): Promise<AuthenticationSession | undefined> {
-		const provider = await this.ensureProvider(providerId);
-
-		const session = await provider.createSession(descriptor);
+		const session = await this.createSession(descriptor);
 		if (session == null) return undefined;
 
-		const key = this.getSecretKey(providerId, provider.getSessionId(descriptor));
+		const key = this.getSecretKey(this.getSessionId(descriptor));
 		await this.container.storage.storeSecret(key, JSON.stringify(session));
 
 		return session;
 	}
 
 	@debug()
 	async getSession(
-		providerId: IntegrationId,
 		descriptor?: IntegrationAuthenticationSessionDescriptor,
 		options?: { createIfNeeded?: boolean; forceNewSession?: boolean },
 	): Promise<ProviderAuthenticationSession | undefined> {
-		if (this.supports(providerId)) {
-			const provider = await this.ensureProvider(providerId);
+		const key = this.getSecretKey(this.getSessionId(descriptor));
 
-			const key = this.getSecretKey(providerId, provider.getSessionId(descriptor));
+		if (options?.forceNewSession) {
+			await this.container.storage.deleteSecret(key);
+		}
 
-			if (options?.forceNewSession) {
+		let storedSession: StoredSession | undefined;
+		try {
+			const sessionJSON = await this.container.storage.getSecret(key);
+			if (sessionJSON) {
+				storedSession = JSON.parse(sessionJSON);
+			}
+		} catch (ex) {
+			try {
 				await this.container.storage.deleteSecret(key);
+			} catch {}
+
+			if (!options?.createIfNeeded) {
+				throw ex;
 			}
+		}
+
+		if (
+			(options?.createIfNeeded && storedSession == null) ||
+			(storedSession?.expiresAt != null && new Date(storedSession.expiresAt).getTime() < Date.now())
+		) {
+			return this.createAndStoreSession(descriptor);
+		}
+
+		return storedSession as ProviderAuthenticationSession | undefined;
+	}
+}
+
+export abstract class LocalIntegrationAuthenticationProvider<
+	ID extends IntegrationId = IntegrationId,
+> extends IntegrationAuthenticationProviderBase<ID> {}
+
+export abstract class CloudIntegrationAuthenticationProvider<
+	ID extends IntegrationId = IntegrationId,
+> extends IntegrationAuthenticationProviderBase<ID> {
+	protected async getBuiltInExistingSession(
+		_?: IntegrationAuthenticationSessionDescriptor,
+	): Promise<ProviderAuthenticationSession | undefined> {
+		return Promise.resolve(undefined);
+	}
+
+	protected override async createSession(
+		descriptor?: IntegrationAuthenticationSessionDescriptor,
+		options?: { authorizeIfNeeded?: boolean },
+	): Promise<ProviderAuthenticationSession | undefined> {
+		const existingSession = await this.getBuiltInExistingSession(descriptor);
+		if (existingSession != null) return existingSession;
+
+		const cloudIntegrations = await this.container.cloudIntegrations;
+		if (cloudIntegrations == null) return undefined;
+
+		let session = await cloudIntegrations.getConnectionSession(this.authProviderId);
+
+		if (session != null && session.expiresIn < 60) {
```

Comment:
We will likely need to make an exception for GitHub here since it comes back with `expiresIn` of 0 but doesn't really expire.

Reply:
true. Good point. I thought about it, but at the last moment have forgotten, because eventually it worked (by refreshing the session on every "createSession" call :-)

Reply:
@axosoft-ramint 

Do we want to report it to the API team? They definitely should return different values for expired and never-expires tokens.

Reply:
btw. done
![image](https://github.com/gitkraken/vscode-gitlens/assets/83944326/141b2310-d2d6-4136-989a-ea7b71cb4b44)